Prometheus Plugin Panic

### System info:

DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.2 LTS"
Telegraf - version 0.13.1

### Steps to reproduce:

1. reload telegraf config
```
2016/06/07 02:31:47 Reloading Telegraf config
2016/06/07 02:31:47 Hang on, flushing any cached metrics before shutdown
2016/06/07 02:31:47 Output [influxdb] buffer fullness: 0 / 1000 metrics. Total gathered metrics: 135674. Total dropped metrics: 0.
2016/06/07 02:31:47 Output [prometheus_client] buffer fullness: 0 / 1000 metrics. Total gathered metrics: 135674. Total dropped metrics: 0.
panic: http: multiple registrations for /metrics

goroutine 1 [running]:
panic(0xe84ec0, 0xc8207fc900)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
net/http.(*ServeMux).Handle(0xc820010ae0, 0x1240ea0, 0x8, 0x7f2b451911b0, 0xc820671770)
        /usr/local/go/src/net/http/server.go:1926 +0x297
net/http.Handle(0x1240ea0, 0x8, 0x7f2b451911b0, 0xc820671770)
        /usr/local/go/src/net/http/server.go:1961 +0x4b
github.com/influxdata/telegraf/plugins/outputs/prometheus_client.(*PrometheusClient).Start(0xc820690660, 0x0, 0x0)
        /home/ubuntu/telegraf-build/src/github.com/influxdata/telegraf/plugins/outputs/prometheus_client/prometheus_client.go:42 +0x7f
github.com/influxdata/telegraf/agent.(*Agent).Connect(0xc82015c548, 0x0, 0x0)
        /home/ubuntu/telegraf-build/src/github.com/influxdata/telegraf/agent/agent.go:53 +0x16a
main.main()
        /home/ubuntu/telegraf-build/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:237 +0x1c76
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -40,6 +40,7 @@ should now look like:
 - [#1399](https://github.com/influxdata/telegraf/issues/1399): Add `read_repairs` statistics to riak plugin.
 - [#1405](https://github.com/influxdata/telegraf/issues/1405): Fix memory/connection leak in prometheus input plugin.
 - [#1378](https://github.com/influxdata/telegraf/issues/1378): Trim BOM from config file for Windows support.
+- [#1339](https://github.com/influxdata/telegraf/issues/1339): Prometheus client output panic on service reload.
 
 ## v1.0 beta 2 [2016-06-21]
 

--- a/plugins/outputs/prometheus_client/prometheus_client.go
+++ b/plugins/outputs/prometheus_client/prometheus_client.go
@@ -25,8 +25,7 @@ var (
 )
 
 type PrometheusClient struct {
-	Listen  string
-	metrics map[string]*prometheus.UntypedVec
+	Listen string
 }
 
 var sampleConfig = `
@@ -35,6 +34,14 @@ var sampleConfig = `
 `
 
 func (p *PrometheusClient) Start() error {
+	defer func() {
+		if r := recover(); r != nil {
+			// recovering from panic here because there is no way to stop a
+			// running http go server except by a kill signal. Since the server
+			// does not stop on SIGHUP, Start() will panic when the process
+			// is reloaded.
+		}
+	}()
 	if p.Listen == "" {
 		p.Listen = "localhost:9126"
 	}
@@ -44,7 +51,6 @@ func (p *PrometheusClient) Start() error {
 		Addr: p.Listen,
 	}
 
-	p.metrics = make(map[string]*prometheus.UntypedVec)
 	go server.ListenAndServe()
 	return nil
 }
@@ -118,24 +124,26 @@ func (p *PrometheusClient) Write(metrics []telegraf.Metric) error {
 				continue
 			}
 
-			// Create a new metric if it hasn't been created yet.
-			if _, ok := p.metrics[mname]; !ok {
-				p.metrics[mname] = prometheus.NewUntypedVec(
-					prometheus.UntypedOpts{
-						Name: mname,
-						Help: "Telegraf collected metric",
-					},
-					labels,
-				)
-				if err := prometheus.Register(p.metrics[mname]); err != nil {
-					log.Printf("prometheus_client: Metric failed to register with prometheus, %s", err)
-					continue
-				}
+			mVec := prometheus.NewUntypedVec(
+				prometheus.UntypedOpts{
+					Name: mname,
+					Help: "Telegraf collected metric",
+				},
+				labels,
+			)
+			collector, err := prometheus.RegisterOrGet(mVec)
+			if err != nil {
+				log.Printf("prometheus_client: Metric failed to register with prometheus, %s", err)
+				continue
+			}
+			mVec, ok := collector.(*prometheus.UntypedVec)
+			if !ok {
+				continue
 			}
 
 			switch val := val.(type) {
 			case int64:
-				m, err := p.metrics[mname].GetMetricWith(l)
+				m, err := mVec.GetMetricWith(l)
 				if err != nil {
 					log.Printf("ERROR Getting metric in Prometheus output, "+
 						"key: %s, labels: %v,\nerr: %s\n",
@@ -144,7 +152,7 @@ func (p *PrometheusClient) Write(metrics []telegraf.Metric) error {
 				}
 				m.Set(float64(val))
 			case float64:
-				m, err := p.metrics[mname].GetMetricWith(l)
+				m, err := mVec.GetMetricWith(l)
 				if err != nil {
 					log.Printf("ERROR Getting metric in Prometheus output, "+
 						"key: %s, labels: %v,\nerr: %s\n",

